In [1]:
import pandas as pd

In [3]:
import numpy as np

In [4]:
import os

In [ ]:
os.chdir('/Users/artemysafronov/Desktop/Учеба/HSE/MachineЛёрнинг/homework/Task1/Pokemon.csv')
df = pd.read_csv('train.csv', encoding='latin-1')

In [5]:
pokemons=pd.read_csv("/Users/artemysafronov/Desktop/MachineЛёрнинг/homework/Task1/Pokemon.csv")

In [7]:
pokemons.head() #first 5 

#                   Name Type 1  Type 2  Total  HP  Attack  Defense  \
0  1              Bulbasaur  Grass  Poison    318  45      49       49   
1  2                Ivysaur  Grass  Poison    405  60      62       63   
2  3               Venusaur  Grass  Poison    525  80      82       83   
3  3  VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4             Charmander   Fire     NaN    309  39      52       43   

   Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0       65       65     45           1      False  
1       80       80     60           1      False  
2      100      100     80           1      False  
3      122      120     80           1      False  
4       60       50     65           1      False

In [8]:
pokemons.describe()

#      Total          HP      Attack     Defense     Sp. Atk  \
count  800.000000  800.00000  800.000000  800.000000  800.000000  800.000000   
mean   362.813750  435.10250   69.258750   79.001250   73.842500   72.820000   
std    208.343798  119.96304   25.534669   32.457366   31.183501   32.722294   
min      1.000000  180.00000    1.000000    5.000000    5.000000   10.000000   
25%    184.750000  330.00000   50.000000   55.000000   50.000000   49.750000   
50%    364.500000  450.00000   65.000000   75.000000   70.000000   65.000000   
75%    539.250000  515.00000   80.000000  100.000000   90.000000   95.000000   
max    721.000000  780.00000  255.000000  190.000000  230.000000  194.000000   

          Sp. Def       Speed  Generation  
count  800.000000  800.000000   800.00000  
mean    71.902500   68.277500     3.32375  
std     27.828916   29.060474     1.66129  
min     20.000000    5.000000     1.00000  
25%     50.000000   45.000000     2.00000  
50%     70.000000   65.000000     3.00000  
75%     90.000000   90.000000     5.00000  
max    230.000000  180.000000     6.00000

In [6]:
pokemons.info() #info about types of variables, names of columns

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 800 entries, 0 to 799
Data columns (total 13 columns):
#             800 non-null int64
Name          800 non-null object
Type 1        800 non-null object
Type 2        414 non-null object
Total         800 non-null int64
HP            800 non-null int64
Attack        800 non-null int64
Defense       800 non-null int64
Sp. Atk       800 non-null int64
Sp. Def       800 non-null int64
Speed         800 non-null int64
Generation    800 non-null int64
Legendary     800 non-null bool
dtypes: bool(1), int64(9), object(3)
memory usage: 75.9+ KB


In [7]:
pokemons.shape

(800, 13)

In [9]:
pokemons.columns

Index(['#', 'Name', 'Type 1', 'Type 2', 'Total', 'HP', 'Attack', 'Defense',
       'Sp. Atk', 'Sp. Def', 'Speed', 'Generation', 'Legendary'],
      dtype='object')

In [12]:
pokemons["HP"].mean() #mean is 69.25875

69.25875

In [13]:
pokemons["HP"].median() #but median is 65, distribution is skewed to the right

65.0

In [16]:
pokemons[pokemons["Legendary"]==True].head() #pokemons which ARE legendary

#                 Name    Type 1    Type 2  Total   HP  Attack  \
156  144             Articuno       Ice    Flying    580   90      85   
157  145               Zapdos  Electric    Flying    580   90      90   
158  146              Moltres      Fire    Flying    580   90     100   
162  150               Mewtwo   Psychic       NaN    680  106     110   
163  150  MewtwoMega Mewtwo X   Psychic  Fighting    780  106     190   

     Defense  Sp. Atk  Sp. Def  Speed  Generation  Legendary  
156      100       95      125     85           1       True  
157       85      125       90    100           1       True  
158       90      125       85     90           1       True  
162       90      154       90    130           1       True  
163      100      154      100    130           1       True

In [17]:
pokemons["Total"].describe() #summary stats for Total

count    800.00000
mean     435.10250
std      119.96304
min      180.00000
25%      330.00000
50%      450.00000
75%      515.00000
max      780.00000
Name: Total, dtype: float64

In [112]:
pokemons["HP"].describe() #summary stats for HP

count    800.000000
mean      69.258750
std       25.534669
min        1.000000
25%       50.000000
50%       65.000000
75%       80.000000
max      255.000000
Name: HP, dtype: float64

In [19]:
pokemons["Attack"].describe() ##summary stats for Attack

count    800.000000
mean      79.001250
std       32.457366
min        5.000000
25%       55.000000
50%       75.000000
75%      100.000000
max      190.000000
Name: Attack, dtype: float64

In [21]:
pokemons["Defense"].describe() #summary stats for Defense

count    800.000000
mean      73.842500
std       31.183501
min        5.000000
25%       50.000000
50%       70.000000
75%       90.000000
max      230.000000
Name: Defense, dtype: float64

In [26]:
pokemons.head(n=800) #looking for last pokemon in the dataset

#                   Name   Type 1  Type 2  Total  HP  Attack  Defense  \
0      1              Bulbasaur    Grass  Poison    318  45      49       49   
1      2                Ivysaur    Grass  Poison    405  60      62       63   
2      3               Venusaur    Grass  Poison    525  80      82       83   
3      3  VenusaurMega Venusaur    Grass  Poison    625  80     100      123   
4      4             Charmander     Fire     NaN    309  39      52       43   
..   ...                    ...      ...     ...    ...  ..     ...      ...   
795  719                Diancie     Rock   Fairy    600  50     100      150   
796  719    DiancieMega Diancie     Rock   Fairy    700  50     160      110   
797  720    HoopaHoopa Confined  Psychic   Ghost    600  80     110       60   
798  720     HoopaHoopa Unbound  Psychic    Dark    680  80     160       60   
799  721              Volcanion     Fire   Water    600  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0         65       65     45           1      False  
1         80       80     60           1      False  
2        100      100     80           1      False  
3        122      120     80           1      False  
4         60       50     65           1      False  
..       ...      ...    ...         ...        ...  
795      100      150     50           6       True  
796      160      110    110           6       True  
797      150      130     70           6       True  
798      170      130     80           6       True  
799      130       90     70           6       True  

[800 rows x 13 columns]

Dataset "Pokemons" contains information about 721 pokemons with their statistics (surprisingly). It has 13 columns which represent: number (#) for each pokemon, name of the pokemon, pokemon's type (type 1), some pokemons also have additional type (type 2), total (is the sum of all numeric stats, except "generation"), health of each pokemon (HP), attack, defence, special attacks (Sp. Atk, for example, "lightning strike), special defence (Sp. Def, how pokemon deals with opponent's special attacks), speed, generation and legendary (that's simple: either legendary or not). Moreover, i did some summary statistics above and below

In [32]:
pokemons.groupby(["Attack","Defense"]).mean() #mean by 2 variables

#  Total     HP  Sp. Atk  Sp. Def  Speed  Generation  \
Attack Defense                                                             
5      5        276.5  335.0  175.0     25.0     85.0   40.0         2.5   
10     10       242.0  540.0  255.0     75.0    135.0   55.0         2.0   
       55       129.0  200.0   20.0     15.0     20.0   80.0         1.0   
       230      213.0  505.0   20.0     10.0    230.0    5.0         2.0   
15     20       349.0  200.0   20.0     10.0     55.0   80.0         3.0   
...               ...    ...    ...      ...      ...    ...         ...   
180    20       386.0  600.0   50.0    180.0     20.0  150.0         3.0   
       100      384.0  780.0  105.0    180.0    100.0  115.0         3.0   
       160      383.0  770.0  100.0    150.0     90.0   90.0         3.0   
185    115      214.0  600.0   80.0     40.0    105.0   75.0         2.0   
190    100      150.0  780.0  106.0    154.0    100.0  130.0         1.0   

                Legendary  
Attack Defense             
5      5              0.0  
10     10             0.0  
       55             0.0  
       230            0.0  
15     20             0.0  
...                   ...  
180    20             1.0  
       100            1.0  
       160            1.0  
185    115            0.0  
190    100            1.0  

[560 rows x 8 columns]

In [35]:
pokemons.groupby("Generation")["Attack"].mean() #mean of Attack sorted by Generation

Generation
1    76.638554
2    72.028302
3    81.625000
4    82.867769
5    82.066667
6    75.804878
Name: Attack, dtype: float64

In [36]:
pokemons.groupby("Generation")["Defense"].mean() #mean of Defense sorted by Generation

Generation
1    70.861446
2    73.386792
3    74.100000
4    78.132231
5    72.327273
6    76.682927
Name: Defense, dtype: float64

In [37]:
pokemons.groupby("Generation")["Attack"].max() #max of Attack sorted by Generation

Generation
1    190
2    185
3    180
4    170
5    170
6    160
Name: Attack, dtype: int64

In [38]:
pokemons.groupby("Generation")["Attack"].min() #min of Attack sorted by Generation

Generation
1     5
2    10
3    15
4     5
5    25
6    22
Name: Attack, dtype: int64

In [40]:
pokemons.sort_values("Speed",ascending=False) #sorting in descending order of column Speed

#                       Name   Type 1  Type 2  Total   HP  Attack  \
431  386          DeoxysSpeed Forme  Psychic     NaN    600   50      95   
315  291                    Ninjask      Bug  Flying    456   61      90   
428  386         DeoxysNormal Forme  Psychic     NaN    600   50     150   
154  142  AerodactylMega Aerodactyl     Rock  Flying    615   80     135   
71    65      AlakazamMega Alakazam  Psychic     NaN    590   55      50   
..   ...                        ...      ...     ...    ...  ...     ...   
658  597                  Ferroseed    Grass   Steel    305   44      50   
486  438                     Bonsly     Rock     NaN    290   50      80   
359  328                   Trapinch   Ground     NaN    290   45     100   
230  213                    Shuckle      Bug    Rock    505   20      10   
495  446                   Munchlax   Normal     NaN    390  135      85   

     Defense  Sp. Atk  Sp. Def  Speed  Generation  Legendary  
431       90       95       90    180           3       True  
315       45       50       50    160           3      False  
428       50      150       50    150           3       True  
154       85       70       95    150           1      False  
71        65      175       95    150           1      False  
..       ...      ...      ...    ...         ...        ...  
658       91       24       86     10           5      False  
486       95       10       45     10           4      False  
359       45       45       45     10           3      False  
230      230       10      230      5           2      False  
495       40       40       85      5           4      False  

[800 rows x 13 columns]

In [114]:
pokemons.sort_values(["Sp. Atk","Sp. Def"],ascending=[True,False]) #sorting in ascending order in column Sp. Atl
#and descending in Sp. Def

#                   Name   Type 1  Type 2  Total   HP  Attack  Defense  \
230  213                Shuckle      Bug    Rock    505   20      10      230   
381  349                 Feebas    Water     NaN    200   20      15       20   
486  438                 Bonsly     Rock     NaN    290   50      80       95   
19    15  BeedrillMega Beedrill      Bug  Poison    495   65     150       40   
488  440                Happiny   Normal     NaN    220  100       5        5   
..   ...                    ...      ...     ...    ...  ...     ...      ...   
71    65  AlakazamMega Alakazam  Psychic     NaN    590   55      50       65   
422  382    KyogrePrimal Kyogre    Water     NaN    770  100     150       90   
426  384  RayquazaMega Rayquaza   Dragon  Flying    780  105     180      100   
429  386     DeoxysAttack Forme  Psychic     NaN    600   50     180       20   
164  150    MewtwoMega Mewtwo Y  Psychic     NaN    780  106     150       70   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
230       10      230      5           2      False  
381       10       55     80           3      False  
486       10       45     10           4      False  
19        15       80    145           1      False  
488       15       65     30           4      False  
..       ...      ...    ...         ...        ...  
71       175       95    150           1      False  
422      180      160     90           3       True  
426      180      100    115           3       True  
429      180       20    150           3       True  
164      194      120    140           1       True  

[800 rows x 13 columns]

Data Indexing

In [66]:
np_pok=np.array(pokemons)

In [67]:
np_pok[2,3]

'Poison'

In [68]:
pokemons.head()

#                   Name Type 1  Type 2  Total  HP  Attack  Defense  \
0  1              Bulbasaur  Grass  Poison    318  45      49       49   
1  2                Ivysaur  Grass  Poison    405  60      62       63   
2  3               Venusaur  Grass  Poison    525  80      82       83   
3  3  VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4             Charmander   Fire     NaN    309  39      52       43   

   Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0       65       65     45           1      False  
1       80       80     60           1      False  
2      100      100     80           1      False  
3      122      120     80           1      False  
4       60       50     65           1      False

In [69]:
np_pok[2,1]

'Venusaur'

In [70]:
np_pok

array([[1, 'Bulbasaur', 'Grass', ..., 45, 1, False],
       [2, 'Ivysaur', 'Grass', ..., 60, 1, False],
       [3, 'Venusaur', 'Grass', ..., 80, 1, False],
       ...,
       [720, 'HoopaHoopa Confined', 'Psychic', ..., 70, 6, True],
       [720, 'HoopaHoopa Unbound', 'Psychic', ..., 80, 6, True],
       [721, 'Volcanion', 'Fire', ..., 70, 6, True]], dtype=object)

Data Slicing

In [73]:
np_pok[1:5,2]

array(['Grass', 'Grass', 'Grass', 'Fire'], dtype=object)

In [79]:
np_pok_copy=np_pok[1:5,3:7].copy()

In [81]:
np_pok_copy[0,0]

'Poison'

In [82]:
np_pok[1:5,3:7]

array([['Poison', 405, 60, 62],
       ['Poison', 525, 80, 82],
       ['Poison', 625, 80, 100],
       [nan, 309, 39, 52]], dtype=object)

In [83]:
np_pok_copy[2,3]

100

In [87]:
np_pok

array([[1, 'Bulbasaur', 'Grass', ..., 45, 1, False],
       [2, 'Ivysaur', 'Grass', ..., 60, 1, False],
       [3, 'Venusaur', 'Grass', ..., 80, 1, False],
       ...,
       [720, 'HoopaHoopa Confined', 'Psychic', ..., 70, 6, True],
       [720, 'HoopaHoopa Unbound', 'Psychic', ..., 80, 6, True],
       [721, 'Volcanion', 'Fire', ..., 70, 6, True]], dtype=object)

Data masking

In [88]:
np_pok[1:5,4:6]<100

array([[False,  True],
       [False,  True],
       [False,  True],
       [False,  True]])

In [92]:
np_pok[0,:]=="Bulbasaur" #searching for Bulbasaur on the first row 

array([False,  True, False, False, False, False, False, False, False,
       False, False, False, False])

Data Fancy indexing

In [96]:
np_pok[3,[2,3,5]] #we are looking for a third row (or forth if we count from 0, still dont know how say right =)) 
                  #and columns 2,3,5

array(['Grass', 'Poison', 80], dtype=object)

In [102]:
np_pok

array([[1, 'Bulbasaur', 'Grass', ..., 45, 1, False],
       [2, 'Ivysaur', 'Grass', ..., 60, 1, False],
       [3, 'Venusaur', 'Grass', ..., 80, 1, False],
       ...,
       [720, 'HoopaHoopa Confined', 'Psychic', ..., 70, 6, True],
       [720, 'HoopaHoopa Unbound', 'Psychic', ..., 80, 6, True],
       [721, 'Volcanion', 'Fire', ..., 70, 6, True]], dtype=object)

In [110]:
row=np.array([3,6,8])  
col=np.array([5,7,10])
np_pok[row,col]

array([80, 78, 100], dtype=object)

In [111]:
np_pok[1:4,[2,3,5]] #fancy indexing + slicing for objects from 1 to 3 rows and columns 2,3,5

array([['Grass', 'Poison', 60],
       ['Grass', 'Poison', 80],
       ['Grass', 'Poison', 80]], dtype=object)